[material-ui] Improve getReactElementRef() utils

sai6855 opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

check https://github.com/mui/material-ui/pull/42818#discussion_r1683583533 for more context

aarongarciah commented on packages/mui-material/src/utils/getChildRef.js:
```diff
@@ -1,8 +1,16 @@
+    if (!React.isValidElement(child)) {
+      console.error(
+        [
+          "MUI: getChildRef doesn't accept array as a child.",
+          'Consider providing a single element instead.',
```
```suggestion
          "MUI: getChildRef expected to receive a single React element child.',
```

aarongarciah commented on packages/mui-material/src/utils/getChildRef.js:
```diff
@@ -1,8 +1,16 @@
-  if (!child || !React.isValidElement(child)) {
-    return null;
+  if (process.env.NODE_ENV !== 'production') {
+    if (!React.isValidElement(child)) {
```
I guess we could use [`React.Children.only`](https://react.dev/reference/react/Children#children-only) although se'd lose the custom error message.

aarongarciah commented on packages/mui-material/src/utils/getChildRef.js:
```diff
@@ -1,8 +1,16 @@
+          'Consider providing a single element instead.',
+        ].join('\n'),
+      );
+      return null;
```
What's the reasoning to return `null` in non-prod envs? I think it's probably better to have a hard error.

oliviertassinari replied:
We can't have code branching between dev and production
```suggestion
```

aarongarciah replied:
Yeah, I meant let the flow continue and let the function error naturally if the param is incorrect.

aarongarciah reviewed (commented):
Left some comments, but let's wait for @oliviertassinari since he suggested the improvements.

oliviertassinari commented on packages/mui-material/src/utils/getChildRef.js:
```diff
@@ -1,8 +1,16 @@
-    return null;
+  if (process.env.NODE_ENV !== 'production') {
+    if (!React.isValidElement(child)) {
+      console.error(
```
I'm missing something. Do we have a concrete example where this warning helping and not harming (creating confusion)?

sai6855 replied:
<img width="953" alt="image" src="https://github.com/user-attachments/assets/c17b16ca-51e3-48c4-a4fa-53942163b78b">
<img width="954" alt="image" src="https://github.com/user-attachments/assets/e5269296-e433-43bf-acf0-96ffbb1dcf3f">

sai6855 replied:
You are right,  `console.error` doesn't provide any help if `child` is invalid react element. Throwing hard error instead of logging error seems to improve DX


when just `console.error` is used
<img width="954" alt="image" src="https://github.com/user-attachments/assets/e5269296-e433-43bf-acf0-96ffbb1dcf3f">

When thowing error instead of console.error

<img width="953" alt="image" src="https://github.com/user-attachments/assets/c17b16ca-51e3-48c4-a4fa-53942163b78b">

oliviertassinari replied:
@sai6855 I think this example

```jsx
<Slide direction="up" in={checked} mountOnEnter unmountOnExit>
  <p>sample</p>
  <p>sample</p>
</Slide>
```

supports the previous argument I was trying to develop: we increase the bundle size for end-users, so we degrade the UX. But on the opposite side, the error is still not actionable, so we don't improve the DX.
I conclude that it makes things worse.

Previously, the experience was (without TypeScript):

<img width="1398" alt="SCR-20240723-bery" src="https://github.com/user-attachments/assets/0c550018-8cc0-4613-8e52-aa80fc51e14f">

which is equally bad. So this looks like a step forward to me. Simpler, less bundle size, faster runtime:

```diff
diff --git a/packages/mui-material/src/utils/getChildRef.js b/packages/mui-material/src/utils/getChildRef.js
index a1d0c2aec5..c24cd926d3 100644
--- a/packages/mui-material/src/utils/getChildRef.js
+++ b/packages/mui-material/src/utils/getChildRef.js
@@ -1,9 +1,4 @@
-import * as React from 'react';
-
 export default function getChildRef(child) {
-  if (!child || !React.isValidElement(child)) {
-    return null;
-  }
   // 'ref' is passed as prop in React 19, whereas 'ref' is directly attached to children in React 18
   // below check is to ensure 'ref' is accessible in both cases
   return child.props.propertyIsEnumerable('ref') ? child.props.ref : child.ref;
```

Now, if we want to avoid the confusion for developers, to not have an error thrown, but only a console error message, I could see us doing:

```diff
diff --git a/packages/mui-material/src/Slide/Slide.js b/packages/mui-material/src/Slide/Slide.js
index d669e811d6..5b3a7a71ed 100644
--- a/packages/mui-material/src/Slide/Slide.js
+++ b/packages/mui-material/src/Slide/Slide.js
@@ -119,6 +119,10 @@ const Slide = React.forwardRef(function Slide(props, ref) {
     ...other
   } = props;

+  // Invalid children provided, bailout
+  if (!React.isValidElement(children)) {
+    return children;
+  }
+
   const childrenRef = React.useRef(null);
   const handleRef = useForkRef(getChildRef(children), childrenRef, ref);

diff --git a/packages/mui-material/src/utils/getChildRef.js b/packages/mui-material/src/utils/getChildRef.js
index a1d0c2aec5..c24cd926d3 100644
--- a/packages/mui-material/src/utils/getChildRef.js
+++ b/packages/mui-material/src/utils/getChildRef.js
@@ -1,9 +1,4 @@
-import * as React from 'react';
-
 export default function getChildRef(child) {
-  if (!child || !React.isValidElement(child)) {
-    return null;
-  }
   // 'ref' is passed as prop in React 19, whereas 'ref' is directly attached to children in React 18
   // below check is to ensure 'ref' is accessible in both cases
   return child.props.propertyIsEnumerable('ref') ? child.props.ref : child.ref;
```

---

Now, there will also be the React 19 prop-types feedback problem for people now using types or using `any`:

<img width="1033" alt="SCR-20240723-bfup" src="https://github.com/user-attachments/assets/a4f2e3c9-561b-4503-a3f3-0fb0adb3bddc">

There, I think that we simply need to create new utils, and do:

```diff
diff --git a/packages/mui-material/src/Slide/Slide.js b/packages/mui-material/src/Slide/Slide.js
index d669e811d6..f316556579 100644
--- a/packages/mui-material/src/Slide/Slide.js
+++ b/packages/mui-material/src/Slide/Slide.js
@@ -1,6 +1,8 @@
 'use client';
 import * as React from 'react';
 import PropTypes from 'prop-types';
+import getDisplayName from '@mui/utils/getDisplayName';
+import checkPropTypes from 'prop-types/checkPropTypes';
 import { Transition } from 'react-transition-group';
 import chainPropTypes from '@mui/utils/chainPropTypes';
 import HTMLElementType from '@mui/utils/HTMLElementType';
@@ -82,11 +84,20 @@ export function setTranslateValue(direction, node, containerProp) {
   }
 }

+function muiCheckPropTypes(Component, props,) {
+  if (process.env.NODE_ENV === 'production') {
+    return;
+  }
+  return checkPropTypes(Component.propTypes, props, 'prop', getDisplayName(Component));
+}
+
 /**
  * The Slide transition is used by the [Drawer](/material-ui/react-drawer/) component.
  * It uses [react-transition-group](https://github.com/reactjs/react-transition-group) internally.
  */
-const Slide = React.forwardRef(function Slide(props, ref) {
+const Slide = React.forwardRef(function SlideIn(props, ref) {
+  muiCheckPropTypes(Slide, props);
+
```

cc @mui/code-infra

Janpot replied:
> This feels like a better DX:

I second that.

> There, I think that we simply need to create new utils

Sure, I perhaps would consider moving the condition around `muiCheckPropTypes(Slide, props);` to shave off a few more bytes and prevent an unnecessary empty function call.

```diff
+const Slide = React.forwardRef(function SlideIn(props, ref) {
+  if (process.env.NODE_ENV !== 'production') {
+    muiCheckPropTypes(Slide, props);
+  }
+
```

Do we have a CI check that verifies certain utilities don't make it in the production bundles?

oliviertassinari replied:
> I perhaps would consider moving the condition around muiCheckPropTypes(Slide, props); to shave off a few more bytes and prevent an unnecessary empty function call.

Fair enough. It's also missing a React.version check to not warn twice in React 18 and lower.

oliviertassinari replied:
Opportunity moved to #43138

oliviertassinari commented:
@sai6855 The changes make sense to me now. We are back to a local maximum 👍.

If we want to get to the next local maximum, I would see:

```diff
diff --git a/packages/mui-material/src/Slide/Slide.js b/packages/mui-material/src/Slide/Slide.js
index d669e811d6..5b3a7a71ed 100644
--- a/packages/mui-material/src/Slide/Slide.js
+++ b/packages/mui-material/src/Slide/Slide.js
@@ -119,6 +119,10 @@ const Slide = React.forwardRef(function Slide(props, ref) {
     ...other
   } = props;

+  // Invalid children provided, bailout
+  if (!React.isValidElement(children)) {
+    return children;
+  }
+
   const childrenRef = React.useRef(null);
   const handleRef = useForkRef(getChildRef(children), childrenRef, ref);
```

+ #43138, but better to be iterative.

aarongarciah commented:
@sai6855 as part of the React 19 effort, @DiegoAndai recently moved `getChildRef` to `mui-utils` under a new name: `getReactNodeRef`. Can you rebase the PR and apply the potential changes there?

sai6855 commented on packages/mui-utils/src/getReactNodeRef/getReactNodeRef.ts:
```diff
@@ -8,13 +8,9 @@ import * as React from 'react';
  * @returns React.Ref<any> | null
  */
 export default function getReactNodeRef(element: React.ReactNode): React.Ref<any> | null {
-  if (!element || !React.isValidElement(element)) {
-    return null;
-  }
-
```
@aarongarciah rebased PR and removed these checks as it was decided here https://github.com/mui/material-ui/pull/43022#discussion_r1687208774

aarongarciah replied:
The comment on [line 15](https://github.com/mui/material-ui/pull/43022/files#diff-2aabece439eee5bb9309cc46b1b3c3b690ee8a7231cbac1ba495cf832e506152R15) needs to be updated since we're not using `React.isValidElement` anymore.

> // We cannot check for it, but isValidElement is true at this point

aarongarciah replied:
Also the JSDoc comment probably needs to be updated:

> It will return null if the node is not a valid React element.

This is not true since the function will error if "non-object" types (which are part of `React.ReactNode`) are passed to it (null, undefined, string, boolean, etc.)

sai6855 commented on packages/mui-utils/src/getReactNodeRef/getReactNodeRef.ts:
```diff
@@ -8,13 +8,9 @@ import * as React from 'react';
   // 'ref' is passed as prop in React 19, whereas 'ref' is directly attached to children in older versions
-  return (element.props as any).propertyIsEnumerable('ref')
-    ? (element.props as any).ref
+  return (element as any)?.props?.propertyIsEnumerable('ref')
```
I'm unsure if type if element should be `React.ReactNode` or `React.ReactElement`, currently this util is used by `Portal` and portal supports multiple children 
https://github.com/mui/material-ui/blob/0137be75949f5a2645ce13272ca8d952dbbc0c6e/packages/mui-base/src/Portal/Portal.tsx#L5

But if type of `element` remains as `React.ReactNode` we need to type case `element` to `any` to make `element.props.propertyIsEnumerable('ref')` work.

Where as if `element` type is `React.ReactElement`, type casting to `any` is not necessary

aarongarciah replied:
Taking into account that:
- The recommended way to use this helper seems to check with `React.isValidElement` first (as per https://github.com/mui/material-ui/pull/43022#discussion_r1687208774)
- `getReactNodeRef`'s behaviour goes like this:
  - returns the ref if the param is an element with a ref
  - returns null for elements without a ref and fragments
  - returns undefined for arrays.
  - throws an error when passing "non-object" children types (null, undefined, booleans, string, etc.)

Maybe it should be named `getReactElementRef` instead of `getReactNodeRef`? And update the param type to be `React.ReactElement` instead of `React.ReactNode`.

Here's a playground showcasing the differences between `getReactElementRef` and `getReactNodeRef`. Note how the former forces consumers to check with `React.isValidElement` before using it, which I think it's a better DX and avoids unexpected errors.

https://codesandbox.io/p/sandbox/getreactnoderef-zjcv2t?file=/src/App.tsx:73,21

Let me know if this makes sense or if I missed something.

cc @DiegoAndai since you worked on this helper, too.

DiegoAndai replied:
The logic behind leaving the `isValidElement` check inside the util is that we were doing the check on most places in which we used the util so it seemed better to incorporate it into it (https://github.com/mui/material-ui/pull/43132#discussion_r1699451479). I'm not against moving the check out, but I would like to understand what's the benefit of doing so. How is forcing consumers to check `isValidElement` better DX?

oliviertassinari replied:
> if type if element should be React.ReactNode or React.ReactElement

+1 for React.ReactElement

> How is forcing consumers to check isValidElement better DX?

@DiegoAndai I tried to demonstrate it in https://github.com/mui/material-ui/pull/43022#discussion_r1687208774. Most of the time, there is already a warning in place to flag when using the API silencing the error up in the the stack makes it easy for developers to notice the problem.

DiegoAndai replied:
Ok, I'm on board with @oliviertassinari and @aarongarciah's proposed changes, lets go with the approach in https://github.com/mui/material-ui/pull/43022#discussion_r1715199139 @sai6855.

oliviertassinari replied:
Sounds great 👍. 

Before we move forward, I think to sync with https://github.com/mui/base-ui/pull/605#discussion_r1759927875 so we move in the same direction.

michaldudak replied:
:+1: We're checking React version and access either `element.ref` or `element.props.ref` based on it. The version check seems to be faster than `propertyIsEnumerable` and IMO it communicates intent better.

sai6855 replied:
```
export default function getReactElementRef(element: React.ReactElement): React.Ref<any> | null {
  const majorVersion = parseInt(React.version.split('.')[0], 10);

  // 'ref' is passed as prop in React 19, whereas 'ref' is directly attached to children in older versions
  if (majorVersion >= 19) {
    return element.props.ref;
  }
  // @ts-expect-error element.ref is not included in the ReactElement type
  // https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/70189
  return element.ref;
}
```

sai6855 commented:
Summary of changes done in this PR

1. Removed `!element || !React.isValidElement(element)` check [here](https://github.com/mui/material-ui/pull/43022/files#diff-959723e81fb330c89eefd3bd7d5335d652b4f1ad0c972e0ea850cea6c442a86cL11), as it was decided it's better DX to move error up to component level [here](https://github.com/mui/material-ui/pull/43022#discussion_r1687208774)
2. Added tests for `getReactElementRef` https://github.com/mui/material-ui/pull/43022/files#diff-55a56a78280026badefee15f47b8823274fe80e1ca3f4a8426c3ffb53d57c1a1 and https://github.com/mui/material-ui/pull/43022/files#diff-54bc148947b0c3f0a3e6872123e20b9615bf2940c1189ce27124597a5233746d
3. Renamed `getReactNodeRef` to `getReactElementRef`
4. Updated usage of `getReactElementRef` in `Portal` component [here](https://github.com/mui/material-ui/pull/43022/files#diff-6a7ffcecb079d43c7d19f218cc19daa9c03fa6129a780194dac0f67e58f824c9R38) and [here](https://github.com/mui/material-ui/pull/43022/files#diff-f7b92ff935ea085e54e49fa813da058d95c167e29d4afffddafee8c1837bc292R38)

aarongarciah reviewed (approved):
Looks good.

I don't have a strong opinion on using `propertyIsEnumerable` vs check the React major. Although I tend to prefer to check features over versions, the version check works well, too. If someone has a strong opinion, speak now 😄

aarongarciah commented on packages/mui-utils/src/getReactNodeRef/getReactNodeRef.ts:
Looks like we lost the git history of this file.

sai6855 replied:
I've deleted the file instead of using `git mv`. do you happen to know any trivial way to transfer git history now, or is it okay if history is lost?

aarongarciah replied:
I'll rebase and push the changes in this PR.

aarongarciah replied:
I rebased and inserted a couple of commits to recover the history. It's fine now.

aarongarciah commented on packages/mui-utils/src/getReactElementRef/getReactElementRef.test.tsx:
```diff
@@ -0,0 +1,39 @@
+import * as React from 'react';
+
+describe('getReactElementRef', () => {
+  it('should return undefined when not used correctly', () => {
```
I wonder if it'd be better to always return `null` in these cases.

sai6855 replied:
No strong opinion, logic might need to change from `element?.ref` to `element?.ref??null` and `element.props?.ref` to `element.props?.ref??null`. If this change doesn't alter bundle size, i'm okay to do the change

aarongarciah replied:
Nah, keep as is. We can always come back to it if needed.

aarongarciah commented:
I'll wait for one last pass from @oliviertassinari before merging since he was the proponent of this change.

oliviertassinari commented on packages/mui-utils/src/getReactElementRef/getReactElementRef.ts:
```diff
@@ -0,0 +1,22 @@
+  const reactMajorVersion = parseInt(React.version.split('.')[0], 10);
+
+  // 'ref' is passed as prop in React 19, whereas 'ref' is directly attached to children in older versions
+  if (reactMajorVersion >= 19) {
```
We could potential save bunch of characters by using string comparison no?

```suggestion
  if (React.version > '19.0.0') {
```

sai6855 replied:
went with parseInt, as js acting weird with string checking if version is greater than 100 😆 .

<img width="437" alt="image" src="https://github.com/user-attachments/assets/eef7f0a7-2656-4000-8556-b23b90a6bafd">

oliviertassinari replied:
> tring checking if version is greater than 100 😆 .

At this point, the logic will be gone, but fair enough. Let's fix this one too: #43820.

oliviertassinari reviewed (commented):
Looks great

aarongarciah commented:
Thank you @sai6855!
